aten/src/ATen/Dispatch_v2.h

```diff
@@ -0,0 +1,170 @@
+//    ... as many types arguments as needed ...
+//  )
+//
+// For example, given an old style:
+//
+//  AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND2(
+//    kComplexHalf,
+//    kHalf,
+//    self.scalar_type(),
+//    "_local_scalar_dense_cpu",
+//    [&] {
+//      scalar_t value = *self.data_ptr<scalar_t>();
+//      r = Scalar(value);
+//    }
+//  )
+//
+// You now write:
+//
+//  AT_DISPATCH_V2(
+//    self.scalar_type(),
+//    "_local_scalar_dense_cpu",
+//    AT_WRAP([&] {
+//      scalar_t value = *self.data_ptr<scalar_t>();
+//      r = Scalar(value);
+//    }),
+//    AT_EXPAND(AT_ALL_TYPES),
+//    AT_EXPAND(AT_COMPLEX_TYPES),
+//    kComplexHalf,
+//    kHalf,
+//  )
+//
+// Notably, it sports the following improvements:
+//
+//  - It is not necessary to specify the arity (e.g.,
+//    AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND{2,3,4,...})
+//    when using the macro
+//
+//  - It is not necessary to specify each dtype individually; if
+//    there is a set of related dtypes and you want to dispatch
+//    over all of them, you can simply say, e.g., AT_EXPAND(AT_INTEGRAL_TYPES)
+//    in your argument list.
+//
+// However, you must remember to wrap the payload body in AT_WRAP, or commas
+// inside your lambda will be improperly handled.  Furthermore, if you more
+// entries to ScalarType than can be supported by this macro, it will fail
+// with an obscure error (due to attempting to concatenate AT_AP with
+// something that is not a number).
+//
+// The implementation strategy is to use the count arguments trick
+// (e.g., as described in https://stackoverflow.com/a/2124385/23845)
+// to discover how many dtypes have been passed, and then dispatch to a
+// hand-written macro for each arity that applies as many DISPATCH_CASE as
+// necessary.  The hand-written macros can be regenerated for other arities
+// with the script below.
+//
+// There is some delicacy in the implementation in controlling when
+// macro expansion occurs, mediated with AT_EXPAND and AT_GUARD.  I mostly
+// relied on GPT4 to help me get it right.
+
+// Public API macros
+
+// See documentation above
+#define AT_DISPATCH_V2(TYPE, NAME, BODY, ...) \
+  AT_DISPATCH_SWITCH(TYPE, NAME, AT_AP_VAR(AT_WRAP(BODY), TYPE, __VA_ARGS__))
+
+// This macro lets you pass an arbitrary expression that may contain internal
+// commas to another macro without having the commas causing the expression
+// to be interpreted as being multiple arguments
+#define AT_WRAP(...) __VA_ARGS__
+
+#define AT_FLOAT8_TYPES \
+  kFloat8_e5m2, kFloat8_e5m2fnuz, kFloat8_e4m3fn, kFloat8_e4m3fnuz
+
+#define AT_INTEGRAL_TYPES kByte, kChar, kInt, kLong, kShort
+#define AT_FLOATING_TYPES kDouble, kFloat
+#define AT_BAREBONES_UNSIGNED_TYPES kUInt16, kUInt32, kUInt64
+#define AT_COMPLEX_TYPES kComplexDouble, kComplexFloat
+#define AT_QINT_TYPES kQInt8, kQUInt8, kQInt32
+// NB: not *actually* all types
+#define AT_ALL_TYPES AT_EXPAND(AT_INTEGRAL_TYPES), AT_EXPAND(AT_FLOATING_TYPES)
```

Comment:
Since we're updating this one, can we rename it? I know it makes it different from the current macro in naming but I think it is worth it if we expect devs will only have to look at this new version going forward.
Maybe AT_BASIC_TYPES or AT_COMMON_TYPES ?

Reply:
I don't intend to port the old call sites, except when I actually have to touch a DISPATCH macro. So the two versions are going to have to coexist for a bit. One thing I guess I am willing to do is do a codemod on both the old and new names so they stay in lockstep. I'd really appreciate the naming lining up exactly.

The naming here is hard too. Basic/Common don't really convey the intention all that well, there is a lot of historical precedent here. AT_TH_TYPES? Lol

Reply:
I like it actually. At least it is clear that it is some arbitrary set and no one will "assume that it contains foo", they will look it up :p 

happy to punt that to a later codemod though if you want them to stay in lockstep.

Reply:
Yeah, I'll do this codemod follow up